### PR TITLE
Shell linters and github action

### LINKED ISSUE #10392: Static analysis github action
The static analysis script we are currently running works well for the majority of cases, however it does need to be run in jenkins along with other integration tests which can take time to queue. Additionally, we could use `golangci-lint` to invoke many more linters and strictly enforce a code style.

This issue will be addressed with numerous PRs. We first need to fix linting errors from the default linters enabled by `golangci-lint`, then those from the extra linters we will enable. Finally we'l set up a github action for static analysis to run separately from the main test suite.

See discussion in https://github.com/lxc/lxd/pull/10384.

### PR DESCRIPTION
**Pull request by markylaing**
Adds and refactors existing shell based linters and adds a github workflow to run static analysis.

Closes #10392

### CODE REVIEW

File: test/lint/client-imports.sh [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#!/bin/sh -eu
+
+echo "Checking for imports that have been added to client or shared..."
+
+OUT=$(go list -f '{{ join .Imports "\n" }}' ./client ./shared/api ./lxc/config | sort -u | grep \\. | diff -u test/godeps.list - || true)
+if [ -n "${OUT}" ]; then
+  echo "ERROR: you added a new dependency to the client or shared; please make sure this is what you want"
+  echo "${OUT}"
+  exit 1
+fi
```
[the remark this answers is not in the document]
Comment (markylaing) [reply test/lint/client-imports.sh:10]: No worries, will double check

**Comment by tomponline** [comment]
./main.sh: 141: test_static_analysis: not found

**Comment by markylaing** [comment]
> ./main.sh: 141: test_static_analysis: not found

@tomponline I've removed this from `test/main.sh` now

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -7,6 +7,41 @@ permissions:
...
+        run: |
+          make update
+
+      - name: Run static analysis
+        run: |
+          export CGO_CFLAGS="${CGO_CFLAGS} -I$(go env GOPATH)/deps/dqlite/include/ -I$(go env GOPATH)/deps/raft/include/"
+          export CGO_LDFLAGS="${CGO_LDFLAGS} -L$(go env GOPATH)/deps/dqlite/.libs/ -L$(go env GOPATH)/deps/raft/.libs/"
+          export LD_LIBRARY_PATH="$(go env GOPATH)/deps/dqlite/.libs/:$(go env GOPATH)/deps/raft/.libs/:${LD_LIBRARY_PATH}"
+          make static-analysis
+
```
Comment (markylaing) [inline .github/workflows/tests.yml:44]: @tomponline @stgraber by this point we have all dependencies to build and unit test LXD (this is required for static analysis anyway). Shall I add the unit tests here and remove the client tests for ubuntu below?

Comment (tomponline) [reply .github/workflows/tests.yml:44]: I'm happy with that, would be more efficient.

File: test/lint/newline-after-block.sh [hunk 3]
```diff
@@ -0,0 +1,17 @@
+#!/bin/sh -eu
+
+echo "Checking that functional blocks are followed by newlines..."
+
+files=$(find ./ -type f -name "*.go")
+
+exit_code=0
+for file in $files
+do
+  RESULT=$(sed -n -e '/^\s*}\s*$/{n;/^\s*\w/{;=}}' "$file" | xargs -L 1 echo "$file" | grep -v 'o$' | sed 's/ /:/g')
```
[the remark this answers is not in the document]
Comment (markylaing) [reply test/lint/newline-after-block.sh]: I've added a comment to explain the oneliner and used `grep -v '\.go'`.

File: .github/workflows/tests.yml (hunk 2 above)
Comment (markylaing) [reply .github/workflows/tests.yml:45]: Done

File: lxc/exec_unix.go [hunk 4]
```diff
@@ -69,74 +71,85 @@ func (c *cmdExec) controlSocketHandler(control *websocket.Conn) {
...
 			}
+
 		case unix.SIGUSR2:
 			logger.Debugf("Received '%s signal', forwarding to executing program.", sig)
 			err := c.forwardSignal(control, unix.SIGUSR2)
 			if err != nil {
 				logger.Debugf("Failed to forward signal '%s'.", unix.SIGUSR2)
 				return
 			}
+
```
Comment (tomponline) [inline lxc/exec_unix.go:144]: This check seems to be going a little too far in my view, as the `case` statements are still part of the overall block. 
What do you think? Should we make an exception here?

Comment (markylaing) [reply lxc/exec_unix.go:144]: I agree that it's taking it a little far. When I initially wrote this check for LXD Cloud I wasn't able to make this exception (I simply found it too difficult, I am sure sed is capable).

**Comment by tomponline** [comment]
@markylaing any ideas why the workflows dont seem to be running on github anymore?

**Comment by markylaing** [comment]
> @markylaing any ideas why the workflows dont seem to be running on github anymore?

@tomponline not a clue :unamused:

File: test/lint/mixed-whitespace.sh [hunk 5]
```diff
@@ -0,0 +1,9 @@
+#!/bin/sh -eu
+
+echo "Checking for mixed tabs and spaces in shell scripts..."
+
+OUT=$(git grep -P '\t' . 2>/dev/null | grep '\.sh:' || true)
```
Comment (simondeziel) [inline test/lint/mixed-whitespace.sh:5]: If the list of `.sh` files containing tabs in them is desired the following sounds simpler:
`git grep -l -P '\t' '*.sh'`

File: test/lint/newline-after-block.sh [hunk 6]
```diff
@@ -0,0 +1,26 @@
+#!/bin/sh -eu
+
+echo "Checking that functional blocks are followed by newlines..."
+
+files=$(find ./ -type f -name "*.go")
```
Comment (simondeziel) [inline test/lint/newline-after-block.sh:5]: `git ls-files '*.go'` is much faster here (on NVMe). Differences worth mentioning: the output produced is ordered differently and the `git` version doesn't have the leading `./`.

File: test/lint/negated-is-bool.sh [hunk 7]
```diff
@@ -0,0 +1,9 @@
+#!/bin/sh -eu
+
+echo "Checking usage of negated shared.Is(True|False)*() functions..."
+
+OUT=$(grep -Pr --exclude-dir=.git '!(shared\.)?Is(True|False).*\(' . 2>/dev/null || true)
```
Comment (simondeziel) [inline test/lint/negated-is-bool.sh:5]: Feels like another candidate for `git grep` too.

File: test/lint/newline-after-block.sh [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+  # This oneliner has a few steps:
+  # 1. sed:
+  #     a. Check for lines that contain a single closing brace (plus whitespace).
+  #     b. Move the pattern space window forward to the next line.
+  #     c. Match lines that start with a word character. This allows for a closing brace on subsequent lines.
+  #     d. Print the line number.
+  # 2. xargs: Print the filename next to the line number of the matches (piped).
+  # 3. If there were no matches, the file name without the line number is printed, use grep to filter it out.
+  # 4. Replace the space with a colon to make a clickable link.
+  RESULT=$(sed -n -e '/^\s*}\s*$/{n;/^\s*\w/{;=}}' "$file" | xargs -L 1 echo "$file" | grep -v '\.go$' | sed 's/ /:/g')
```
Comment (simondeziel) [inline test/lint/newline-after-block.sh:19]: Is the first address selector allowing a trailing white space after the closing `}`?

Thanks for the comment it helps but I still can't decypher it fully. I ran it on `main` and it didn't catch anything despite having manually removed a `\n` after a function block.

File: test/lint/no-short-form-imports.sh [hunk 9]
```diff
@@ -0,0 +1,9 @@
+#!/bin/sh -eu
+
+echo "Checking for short form imports..."
+
+OUT=$(grep -r -n --include \*.go -P '^\s*import\s+"' . 2>/dev/null | grep -v '"C"')
```
Comment (simondeziel) [inline test/lint/no-short-form-imports.sh:5]: `git grep` here too please.

File: test/lint/trailing-space.sh [hunk 10]
```diff
@@ -0,0 +1,9 @@
+#!/bin/sh -eu
+
+echo "Checking that there are no trailing spaces in shell scripts..."
+
+OUT=$(grep -r --exclude-dir=.git " $" . 2>/dev/null | grep '\.sh:' || true)
```
Comment (simondeziel) [inline test/lint/trailing-space.sh:5]: `git grep` here too please.

File: test/lint/newline-after-block.sh (hunk 6 above)
Comment (tomponline) [reply test/lint/newline-after-block.sh:5]: Hrm, this wouldn't work if manually running `make static-analysis` on uncommited files though right?

File: test/lint/newline-after-block.sh (hunk 8 above)
Comment (markylaing) [reply test/lint/newline-after-block.sh:19]: @simondeziel yes the first selector is finding any line which contains *only* a single `}` and whitespace. Then, moving the pattern space forward to the next line, we have a match if it begins with a word character.

This should catch any statements that are run on lines immediately following a closing brace:
```golang
if err != nil {
    return err
}
x = 10
```

**Pull request merged**